uptane/common.py

```diff
@@ -62,21 +90,136 @@ def sign_signable(signable, keys_to_sign_with):
 
     # Else, all is well. Sign the signable with the given key, adding that
     # signature to the signatures list in the signable.
-    signable['signatures'].append(
-        tuf.keys.create_signature(
-        signing_key, signable['signed'], force_treat_as_pydict=True))
+    signable['signatures'].append(sign_over_metadata(
+        signing_key, signable['signed'], datatype=datatype,
+        metadata_format=metadata_format))
+
+
+  uptane.formats.ANY_SIGNABLE_UPTANE_METADATA_SCHEMA.check_match(signable)
+
+  return
+
+
+
+
+
+def sign_over_metadata(
+    key_dict, data, datatype, metadata_format=tuf.conf.METADATA_FORMAT):
+  """
+  Almost exactly identical to the function simultaneously added to TUF,
+  tuf.sig.sign_over_metadata(). Requires datatype.
+  Must differ in Uptane simply because it is not possible to convert
+  Uptane-specific metadata (Time Attestations, ECU Manifests, and Vehicle
+  Manifests) to or from ASN.1/DER without knowing which of those three
+  types of metadata you're dealign with, and this conversion is required for
+  signing and verifying signatures.
+
+  Higher level function that wraps tuf.keys.create_signature, and works
+  specifically with Time Attestations, ECU Manifsts, and Vehicle Manifests that
+  will be in JSON or ASN.1/DER format.
+
+  See tuf.keys.create_signature for overall functionality and the arguments
+  key_dict and data.
+
+  Optional argument:
+
+    metadata_format: (optional; default based on tuf.conf.METADATA_FORMAT)
+
+      If 'json', treats data as a JSON-friendly Python dictionary to be turned
+      into a canonical JSON string and then encoded as utf-8 before signing.
+      When operating TUF with DER metadata but checking the signature on some
+      piece of JSON for some reason, this should be manually set to 'json'. The
+      purpose of this canonicalization is to produce repeatable signatures
+      across different platforms and Python key dictionaries (avoiding things
+      like different signatures over the same dictionary).
+
+      If 'der', the data will be converted into ASN.1, encoded as DER,
+      and hashed. The signature is then checked against that hash.
+  """
+
+  tuf.formats.ANYKEY_SCHEMA.check_match(key_dict)
+  # TODO: Check format of data, based on metadata_format.
+  # TODO: Consider checking metadata_format redundantly. It's checked below.
+
+  if metadata_format == 'json':
+    data = tuf.formats.encode_canonical(data).encode('utf-8')
+
+  elif metadata_format == 'der':
+
+    # TODO: Have convert_signed_metadata_to_der take just the 'signed' element
+    # so we don't have to do this silly wrapping in an empty signable.
+    data = asn1_codec.convert_signed_metadata_to_der(
+        {'signed': data, 'signatures': []}, only_signed=True, datatype=datatype)
+    data = hashlib.sha256(data).digest()
```

Comment:
An attempt was made in the past to make hashing algorithms configurable.  Although at the time it didn't seem doable, due to prepended sha512 digests and filename length restrictions on some systems, a later change to the specification allowed for configurable hashing algorithms in all of TUF.  The change was that version numbers are now prepended to all metadata filenames, rather than digests.

After version numbers were added to filenames, Root and Snapshot were the only metadata required to have digests prepended, but they too have now been modified to only prepend version numbers.  In addition, multiple hashing algorithms are supported for KEYIDs

Presently, I think the hashing algorithm for digest can be made configurable.